### ISSUE TITLE
fallbacks for techniques

### ISSUE THREAD

**Issue by RemiArnaud** [KhronosGroup/glTF#84]
According to the space, the following technique names can be provided to provide more guidance:

Lambert
Phong
Blinn-Phong

1) missing from that list: constant

2) the converter should follow the spec, specifically for duck.json, 'Blinn-Phong' should be used instead of 'technique1'

3) regarding the spec, this is not a name, this is an id.

**Comment by pjcozzi** [comment]
As far as I can tell, this is duplicate with #361 (let me know if someone thinks otherwise), and we all agree with @tparisi's comment above:

> Here is my view: a valid glTF material could have any of the following.

> 1. No shader code supplied - just technique info and param values: requires
application to have an implementation of the technique;
2. Shader code is  supplied with the content, along with technique info;
the application can use the supplied shader code or its own implementation
if it has one;
3. Shader only is supplied, no common technique information; the
application must use the shader code supplied.

> Converters and exporters can provide options for what to output.

**Comment by RemiArnaud** [comment]
What's the size difference between 2 and 3 ?
is it worth the added complexity? (I'd remove option 3 to simplify design/loader)

**Comment by pjcozzi** [comment]
Removing 3 will depend on how #361 comes out.  If the materials are described in an extension, we wouldn't be able to remove 3 since then it would mean that glTF requires the extension.  I also don't think we should remove 3 since a model could contain shaders that can't be represented by the fixed-function materials, and the converter would have the burden of creating some type of material even if the engine doesn't use it, which is the case in Cesium.

**Comment by RemiArnaud** [comment]
3) prevent simple applications to take a peak at the model. 

I was not proposing that the shader provided and the information provided in the technique need to provide same result. Its in fact not possible in the general case, so why bother?

**Comment by pjcozzi** [comment]
> 3) prevent simple applications to take a peak at the model.

Not sure what you mean.

> I was not proposing that the shader provided and the information provided in the technique need to provide same result.

It would still be burdensome to have to generate anything at all when all the runtime wants is the shader.

**Comment by RemiArnaud** [comment]
ok, so what about just 1) then ?

**Comment by pjcozzi** [comment]
Just 3 prevents engines from generating their own shaders at runtime like Three.js does, see #361.

**Comment by RemiArnaud** [comment]
I'm confused. 

If you want engines to generate their own sharers, then we're back to my original post, do not offer 3) always provide 1) or 2)

**Comment by pjcozzi** [comment]
We are repeating a discussion from years ago - read this issue from the top.  We want the flexibility to use shaders in the glTF _or_ to generate them on the client.  This depends on the needs of the client.

**Comment by RemiArnaud** [comment]
Ok, this implies that you know the needs of the client when you generate a glTF file - which I'm perfectly fine with. 
In that case we need 1 and 3, not 2

**Comment by pjcozzi** [comment]
I am also OK with dropping 2, but am OK either way.  @tparisi do you have a need for this:

> 2 . Shader code is supplied with the content, along with technique info; the application can use the supplied shader code or its own implementation if it has one;

**Comment by pjcozzi** [comment]
As discussed offline, @tparisi we need to converge on this issue for the draft 1.0 spec.  Above, you said:

> Here is my view: a valid glTF material could have any of the following.

> 1. No shader code supplied - just technique info and param values: requires
application to have an implementation of the technique;
2. Shader code is  supplied with the content, along with technique info;
the application can use the supplied shader code or its own implementation
if it has one;
3. Shader only is supplied, no common technique information; the
application must use the shader code supplied.

> Converters and exporters can provide options for what to output.

For conformance, we could:
* Only allow (2) to guarantee glTF assets are interoperable.  This puts the burden on the converter to output both even if the technique info is just Lambert.
* Only allow (1) or (3).  This means the converter and runtime only need to support one approach, but it can fragment the glTF ecosystem, e.g., if a new OBJ-to-glTF converter only generated only the technique info and Cesium only supported shaders.

I _think_ the best approach is to either

a) Require both and let the runtime decide which to use (this puts the burden on the converter, which is in the spirit of glTF, and only adds trivial overhead)
b) Require only the shader, and use a (perhaps widely supported) extension for materials ( #361)

**Comment by mlimper** [comment]
I believe it would be ideal if there would be a way to save the overhead of having shader code in all cases, for various reasons.

One reason: Specifying a shader with all its parameters is nice to get a portable appearance - but then, what happens if your lighting is not specified in the glTF file, but - for example - in an X3D scene or inside the application code? In some cases, one might want to use an own shader and lighting setup (many lights instead of one, point lights instead of directional ones, and so on), and one might want to just somehow inject the material information, without necessarily having to use the lights, projection matrix etc. from the glTF scene.

Another reason would be that, given that I only want to use a very simple material, I would waste a lot of space when specifying a full technique, program, shader, ... for each (binary) glTF file.

Finally, another reason, maybe even the most important one: Assuming a very simple default material, the overhead of supporting this material in a WebGL implementation (for example) is much lower than supporting the whole technique-program-shader-passes-model would be.

If it would be possible to use only a material definition and refer to a standard technique (which could also be linked) that would be a very nice solution... maybe there could be a urn for this (cp. [web3d proposal in rfc3541](https://www.ietf.org/rfc/rfc3541.txt)):
```
"materials": {
        "blinn3-fx": {
            "instanceTechnique": {
                "technique": "urn:khronos:gltf:materials:blinn",
                "values": {
                    "ambient": [0,0,0,1],
                    "diffuse": [0.7,0.7,0.7,1],
                    "emission": [0,0,0,1],
                    "shininess": 38.4,
                    "specular": [0.2,0.2,0.2,1]
                }
            },
            "name": "blinn3"
        }
    }
```
(Maybe one could also use one of the COLLADA materials?)

**Comment by tparisi** [comment]
@mlimper yes I have exactly the same opinion. 

1. There is a ton of content out there that either uses the common techniques like Blinn and Lambert, or doesn't care about shading.

2. Importing glTF data into an application that has external lighting defined in the app means that there are potential mismatches in the lighting if we go with the shader in the content vs. just using the material information. Hmm yes we probably want to separate out the material parameters from the shading technique itself so you have the type of interoperability you are talking about. 

Related to this are #93 and #409 ... there had been a proposal to do away with lights completely and just "do it all in the shader." I think this way lies madness. And I think this is the crux of the issue. On the one hand, shaders give the author full control over presentation and minimize the impact on the loader runtime; OTOH it means that the imported content is a black box and potentially useless for some applications. 

I think, a lot of times you just want to know this is a Phong material already. I know, that's a "fixed pipeline" concept but I think we're in danger of throwing the baby out with the bathwater. Just because fixed function was super-limiting a decade and a half ago doesn't mean that it's all bad.

Wondering if we need to design some metadata or profile info that indicates this stuff, e.g. denotes a lighting model or set of them, and an associated materials system. If none of that is there, it's all raw shader-based. If no shaders are there, the app does whatever it wants. This could work but it makes me wonder if it's going to make conformance hard and whether we're setting up for COLLADA-style fragmentation.

**Comment by RemiArnaud** [comment]
This is an area where we did not see fragmentation in COLLADA, because we have mandatory common description (this is phong) and optional shader. 
This work quite well for making sure all COLLADA application can at least display the models with basic shaders. 
So we're back to making gltf mandatory provide shader description and optional shader programs.

**Comment by tparisi** [comment]
@RemiArnaud I couldn't parse this sentence, what do you mean:

"So we're back to making gltf mandatory provide shader description and optional shader programs."

**Comment by RemiArnaud** [comment]
Meaning havig shader description is mandatory. And having shader program is optional. 
So 1) and 2)

**Comment by tparisi** [comment]
For common techniques I see no problem with this...

**Comment by pjcozzi** [comment]
Given the interop and simplicity interests, I am now in favor of this.  My main concern is how fast can we define the schema **and** have a converter and loader implementation that proves it?

@mlimper do you have the bandwidth to propose a schema?  I like your direction of not requiring techniques (and therefore `shader` and `program` objects), and instead allowing materials to reference a well know technique (Blinn, etc.) almost like parameter semantics.  We'll also want to design this so it extends well to PBR and facilities glTF extensions for new materials.

As for conformance, perhaps all assets will include the "fixed-function" material, which runtimes need to support (**but generating a shader for all the light combinations will be painful, perhaps the engine is free to use their own light and ignore the glTF lights for conformance?**).  If an asset has a shader material, the runtime is free to use it instead, but not required.

I also suggest we remove `passes` from `technique` since glTF 1.0 only has one pass.  It is also not clear to me why `material` has a `instanceTechnique`, when those properties could just be part of the material.

**Comment by mlimper** [comment]
> @mlimper do you have the bandwidth to propose a schema?

Unfortunately I think I won't find the time to work on something - but could it be possible to refer to an existing COLLADA or VRML/X3D material definition? In the X3D spec, for example, it is very well-specified how the lighting computations / material model should be implemented... maybe something similar exists for COLLADA?

**Comment by pjcozzi** [comment]
Yes, COLLADA has something similar.  We can use this as inspiration, but we still need to nail down the schema, semantics, etc.

**Comment by pjcozzi** [comment]
Also, does that change your mind about having time? :)

**Comment by tparisi** [comment]
Sorry I'm confused. We already have the fixed function technique... in the form of the pass->details->common profile thing. Do you think it's under-specified or broken enough to warrant changing it at this late date?

You and I talked about passes the other day. I thought we agreed that it's a good idea for forward compatibility, though of course we can't be sure until we try it. On that note I would suggest changing `pass` from a single string value to an array.

**Comment by pjcozzi** [comment]
>  Do you think it's under-specified or broken enough to warrant changing it at this late date?

Yes, for sure.  We are talking about dropping the need to require `techniques` since "Blinn", for example, _is the technique_.

>  I thought we agreed that it's a good idea for forward compatibility, though of course we can't be sure until we try it.

We really don't know what multi-pass will look like so there's a good chance we'll be wrong here.  We also don't do "forward compatibility" design anywhere else in the schema.  Techniques and materials appear more complicated than they need to be.

**Comment by tparisi** [comment]
Well amen to that; I've felt that way all along.

**Comment by mlimper** [comment]
Just discussed this with @mthoener, actually we feel that it could even be the other way around:

glTF could provide a very small number of built-in techniques, such as the fixed function technique and  also a PBR technique, because this will probably be sufficient for very most applications.
On top of that, for the small amount of applications needing a custom shader, implementations could decide to support this feature, which would then be an extension. This would make it much easier to implement a loader that covers the core glTF functionality.

**Comment by tparisi** [comment]
Well, as much as I agree with you about the relative priorities, I know that some implementations like Cesium just use the shader. Right @pjcozzi

**Comment by fabrobinet** [comment]
Please don't make glTF a rendering specification. If you need conventions to describe lighting models you can relate them to COLLADA profiles or whatever other lighting scheme. For instance "Blinn" shouldn't be part of the core spec but can be mentioned as part of a lighting model description matching COLLADA specification inside a extra or extension ? (just a matter to move "details" property in the right place)  that's the bit I am not sure, but the bottom line is to keep glTF generic and not tied to lighting models because GL isn't. (anymore).

**Comment by fabrobinet** [comment]
I meant not a lighting model specification.

**Comment by pjcozzi** [comment]
@fabrobinet from a purist perspective, I agree, but pragmatically. this is such a commonly requested feature, maps well to most engine materials, and really simplifies the glTF asset for many use cases.  @tparisi will have a schema proposal soon.  It would be great if you have time to provide feedback.

**Comment by mlimper** [comment]
Could we have a very simple example that shows how to use a very simple material model?
I just took what is there for the glTF version of the collada duck - here's how the technique currently looks like:
```
"technique1": {
            "parameters": {
                "ambient": {
                    "type": 35666
                },
                "diffuse": {
                    "type": 35678
                },
                "emission": {
                    "type": 35666
                },
                "light0Color": {
                    "type": 35665,
                    "value": [
                        1,
                        1,
                        1
                    ]
                },
                "light0Transform": {
                    "semantic": "MODELVIEW",                    
                    "type": 35676
                },
                "modelViewMatrix": {
                    "semantic": "MODELVIEW",
                    "type": 35676
                },
                "normal": {
                    "semantic": "NORMAL",
                    "type": 35665
                },
                "normalMatrix": {
                    "semantic": "MODELVIEWINVERSETRANSPOSE",
                    "type": 35675
                },
                "position": {
                    "semantic": "POSITION",
                    "type": 35665
                },
                "projectionMatrix": {
                    "semantic": "PROJECTION",
                    "type": 35676
                },
                "shininess": {
                    "type": 5126
                },
                "specular": {
                    "type": 35666
                },
                "texcoord0": {
                    "semantic": "TEXCOORD_0",
                    "type": 35664
                }
            },
            "pass": "defaultPass",
            "passes": {
                "defaultPass": {
                    "details": {
                        "commonProfile": {
                            "extras": {
                                "doubleSided": false
                            },
                            "lightingModel": "Blinn",
                            "parameters": [
                                "ambient",
                                "diffuse",
                                "emission",
                                "light0Color",
                                "light0Transform",
                                "modelViewMatrix",
                                "normalMatrix",
                                "projectionMatrix",
                                "shininess",
                                "specular"
                            ],
                            "texcoordBindings": {
                                "diffuse": "TEXCOORD_0"
                            }
                        },
                        "type": "COLLADA-1.4.1/commonProfile"
                    },
                    "instanceProgram": {
                        "attributes": {
                            "a_normal": "normal",
                            "a_position": "position",
                            "a_texcoord0": "texcoord0"
                        },
                        "program": "program_0",
                        "uniforms": {
                            "u_ambient": "ambient",
                            "u_diffuse": "diffuse",
                            "u_emission": "emission",
                            "u_light0Color": "light0Color",
                            "u_light0Transform": "light0Transform",
                            "u_modelViewMatrix": "modelViewMatrix",
                            "u_normalMatrix": "normalMatrix",
                            "u_projectionMatrix": "projectionMatrix",
                            "u_shininess": "shininess",
                            "u_specular": "specular"
                        }
                    },
                    "states": {
                        "enable": [
                            2884,
                            2929
                        ]
                    }
                }
            }
        }
```
That's just too much, and also too specific...

Can't it be possible to refer to a common technique instead, using a urn or something like that?

(If already there's another, more simple way to achieve a blinn/phong appearance with less lines of JSON, I would be happy to see it. Maybe the example is just bloated somehow?)

**Comment by pjcozzi** [comment]
> Can't it be possible to refer to a common technique instead, using a urn or something like that?

Yes, @tparisi is working on a schema proposal.  Watch this issue.

**Comment by tparisi** [comment]
Look, let's get real about this. 

The only thing a technique needs to specify for a Blinn shader is the enum "Blinn." Look at what is currently being generated for the parameters for the Blinn shader, all of which are being ignore e.g. in the Three.js implementation:


```
                                "light0Color",
                                "light1Color",
                                "light1ConstantAttenuation",
                                "light1LinearAttenuation",
                                "light1QuadraticAttenuation",
                                "light1Transform",
                                "light2Color",
                                "light2ConstantAttenuation",
                                "light2LinearAttenuation",
                                "light2QuadraticAttenuation",
                                "light2Transform",
                                "modelViewMatrix",
                                "normalMatrix",
                                "projectionMatrix",
```

For a built-in Blinn shader, all of the above information is inferred by the runtime and passed to built-in shaders. So we should, at a minimum, remove this bit of waste.

I understand's Fabrice's "purist" viewpoint as Patrick called it, and I appreciate it. But in addition to the pragmatic concerns, there is also the technical side of this: the current schema is WAY too verbose and complex to define the way we are doing it. We can make that better.

I like Max's dictionary/URI idea so that we could look up known shaders in a registry but that doesn't really solve the problem either. I don't think we can just describe the common techniques as another generic GLSL based technique that we fetch programs from a server for, without an engine having some built-in knowledge; this because we also have to take lights into account. Related to #409, we don't have away to even specify the direction for a directional or spot light, within the node itself, which is where it needs to be. Right now the current shaders in the duck model are calculating a light direction based on the light's model position; but a directional light has a direction independent of that. To do this correctly, I believe we would need to be able to extend a node to have other properties, such as a defined direction vector (that's assuming we would ditch lights altogether, which is where this thinking is headed...)

Give me a little time and I'll try to come up with a much more concise way to define a common technique.

**Comment by tparisi** [comment]
Another idea here... 

1. take Max's URN idea, no need to downlload them from anywhere, it's just a name tag
2. define a `fallbackTechnique` property - this is where the URN for common shaders would go
3. keep `technique` the way it is (though hopefully simplify the instances and pass part as we are discussing)

The only issue that Patrick brought up is about the values. Currently the values have to map to parameter names in the technique pass, but with this proposal we'd have two different techniques. Any suggestions for how to deal with that?

**Comment by tparisi** [comment]
So first, let's do a proposal for the fallbacks based on this:

```josn
    "materials": {
            "blinn3-fx": {
                "technique": "technique1",
                "fallback": "urn:khronos:gltf:materials:blinn",
                "values": {
                    "ambient": [0,0,0,1],
                    "diffuse": [0.7,0.7,0.7,1],
                    "emission": [0,0,0,1],
                    "shininess": 38.4,
                    "specular": [0.2,0.2,0.2,1]
                }
                "name": "blinn3"
            }
        }
```

**Comment by tparisi** [comment]
Now, here's a proposal based on Patrick's urging to simplify things by getting rid of multiple passes. It also gets rid of the "details" part because that is taken care of in the common technique. Nit: we need to put the `doubleSided: false` extra somewhere

```json

    "techniques": {
        "technique1": {
            "parameters": {
                "ambient": {
                    "type": 35666
                },
                "diffuse": {
                    "type": 35678
                },
                "emission": {
                    "type": 35666
                },
                "light0Color": {
                    "type": 35665,
                    "value": [
                        1,
                        1,
                        1
                    ]
                },
                "light0Transform": {
                    "semantic": "MODELVIEW",
                    "source": "directionalLight1",
                    "type": 35676
                },
                "modelViewMatrix": {
                    "semantic": "MODELVIEW",
                    "type": 35676
                },
                "normal": {
                    "semantic": "NORMAL",
                    "type": 35665
                },
                "normalMatrix": {
                    "semantic": "MODELVIEWINVERSETRANSPOSE",
                    "type": 35675
                },
                "position": {
                    "semantic": "POSITION",
                    "type": 35665
                },
                "projectionMatrix": {
                    "semantic": "PROJECTION",
                    "type": 35676
                },
                "shininess": {
                    "type": 5126
                },
                "specular": {
                    "type": 35666
                },
                "texcoord0": {
                    "semantic": "TEXCOORD_0",
                    "type": 35664
                }
            },
            "attributes": {
                "a_normal": "normal",
                "a_position": "position",
                "a_texcoord0": "texcoord0"
            },
            "program": "program_0",
            "uniforms": {
                "u_ambient": "ambient",
                "u_diffuse": "diffuse",
                "u_emission": "emission",
                "u_light0Color": "light0Color",
                "u_light0Transform": "light0Transform",
                "u_modelViewMatrix": "modelViewMatrix",
                "u_normalMatrix": "normalMatrix",
                "u_projectionMatrix": "projectionMatrix",
                "u_shininess": "shininess",
                "u_specular": "specular"
            }
            "states": {
                "enable": [
                    2884,
                    2929
                ]
            }
        }
    }
```

**Comment by tparisi** [comment]
@pjcozzi that's my proposal, a straw man anyway. please let me know

**Comment by pjcozzi** [comment]
> Now, here's a proposal based on Patrick's urging to simplify things by getting rid of multiple passes...

This is spot on with what I was thinking.

**Comment by tparisi** [comment]
Cool; now when we say "fallback" do we really mean that. As in, an implementation should try to render with the shader, but use the fallback if it can't? Or do we mean, use the common technique if you have but go with the shader if you don't? In which case I think the shader is the "fallback."

**Comment by pjcozzi** [comment]
That feedback is just for your change to techniques.  I am working on the material fallback now.  Give me a few minutes please...

**Comment by pjcozzi** [comment]
I would like to use the same values for both the technique and the fallback, but they could be completely different., and the fallback's values might not be a subset of the technique's values.

Perhaps:

```javascript
    "materials": {
            "material-id": {
                "technique": {
                    "name" : "technique1",
                    "values": {
                        "diffuse": [0.7,0.7,0.7,1],
                        "shininess": 38.4,
                        "specular": [0.2,0.2,0.2,1]
                    }
                }
                "fallback": {
                    "name" : "LAMBERT", // naming convention similiar to glTF semantics
                    "values": {
                        "diffuse": [0.5,0.5,0.5,1]
                    }
                }
                "name": "my-material"
            }
        }
```

**Comment by tparisi** [comment]
or use `technique` instead of `name`, would be more consistent with everything else we're doing, no?

**Comment by pjcozzi** [comment]
> or use technique instead of name, would be more consistent with everything else we're doing, no?

Ah, yes, you are right.  It isn't even a name; it is an id.

> Cool; now when we say "fallback" do we really mean that. As in, an implementation should try to render with the shader, but use the fallback if it can't? Or do we mean, use the common technique if you have but go with the shader if you don't? In which case I think the shader is the "fallback."

I _think_ fallback should be something like `common` and only it should be required (so a glTF asset doesn't need any top-level `techniques`, `programs`, or `shaders`.  If `technique` is provided, then it is optional to use, a "fall forward" as you called it.

The spec can also explicitly define the `values` for each `fallback`, e.g., `LAMBERT` has only `diffuse`, which is an array of four floats (and so on).  The spec will also need to define that the `primitive` will provide the right accessors, e.g., one that contains attributes with `POSITION` and `NORMAL` semantics.  We also need to define textures (since this lacks the technique and details).  Anyway, it's possible that most of this can just be defined by the spec instead of having a really complicated schema, but we need to think through it and play with some pseudo code.

**Comment by tparisi** [comment]
So let's try this:

```javascript
    "materials": {
            "material-id": {
                "technique": {
                    "technique" : "technique1",
                    "values": {
                        "diffuse": [0.7,0.7,0.7,1],
                        "shininess": 38.4,
                        "specular": [0.2,0.2,0.2,1]
                    }
                }
                "common": {
                    "technique" : "LAMBERT", // naming convention similiar to glTF semantics
                    "values": {
                        "diffuse": [0.5,0.5,0.5,1]
                    }
                }
                "name": "my-material"
            }
        }
```

**Comment by pjcozzi** [comment]
Yes, this has potential, now we need to think through if the spec can define the `values` for each common material, the attribute inputs, the uniform inputs (e.g., model matrix semantic), and how textures are handled.  I _think_ this can all be well defined in the spec as opposed to a big schema.

**Comment by tparisi** [comment]
I do think we can. Blinn, Phong, Lambert and Unlit are my votes and I think they are all well-defined and easy to specify in this day and age. It's probably a few hours' work to write it up, but let's make another issue for that.

BTW I also think this opens the material more easily to adding an `extensions` property that would have completely separate techniques and values... :-)

**Comment by pjcozzi** [comment]
I agree with your list of Blinn, Phong, Lambert and Unlit.  However, I think we need to consider them in this issue since it could influence the spec, e.g., how are diffuse textures handled, is a texture allowed to be a specular map, etc.  We should keep it really simple and let users define new ones for their very specific needs.

Perhaps the base from COLLADA is a good start: https://www.khronos.org/files/collada_spec_1_5.pdf

For example, see Page 275.

**Comment by tparisi** [comment]
Yes the COLLADA spec basically defines all these, though with different naming conventions and such e.g. fx_common_color_or_texture_type. I would go with what's in COLLADA, these are good enough.

**Comment by tparisi** [comment]
Actually can we normatively reference the COLLADA spec, the way we are already doing for WebGL with enums and types etc.?

Then we don't have to duplicate all that prose...

**Comment by pjcozzi** [comment]
I dunno...I would rather be explicit here - and we also want to be clear, from a positioning perspective, that glTF is not related to COLLADA.

Can you draft up an example for Lambert to aid the discussion?

**Comment by tparisi** [comment]
Will do; not today. Fine we don't have to reference, we will write our own.

**Comment by pjcozzi** [comment]
@tparisi any update?

**Comment by tparisi** [comment]
Flat out w day job today. I'll have something over the weekend 

Tony Parisi
CTO at Large
415.902.8002

**Comment by tparisi** [comment]
Still a lot to think about here...  I think we need to nail down conformance requirements before we can finalize the design. 

I'm starting to agree with Fabrice that we don't want to *mandate* a lighting model. This argues for the minimum conformance being to support the user-defined techniques.

OTOH it would be *very* nice to be able to support content that has no user-defined shading in it at all, if e.g. the model is intended to be unlit or all Phong or whatever... which argues for the minimum conformance level being to support common techniques.

I see both of the above scenarios being really useful and we need to support both. Any inspirations on how to get through this?

**Comment by mlimper** [comment]
> OTOH it would be very nice to be able to support content that has no user-defined shading in it at all, if e.g. the model is intended to be unlit or all Phong or whatever... which argues for the minimum conformance level being to support common techniques.

I would guess that, in practice, it should really be no problem to ask for blinn-phong lighting being supported by every implementation. On the other hand, handling shader code or even shader descriptions definitely means some more overhead to take care of, not mentioning the conceptual problems already discussed (for example, which / how many lights do we use? do we have to use GLSL ES, or can the asset also be rendered with Desktop OpenGL using GLSL?)

It will probably be sufficient to ask for support of only one lighting model (plus constant). The one implemented in the standard OpenGL fixed function pipeline (blinn-phong AFAIK?) seems like a good candidate, because it is very well-known and well-specified. Shouldn't be too much work to refer to it, hopefully.
Also, older assets, or such using the common ISO standard formats (like VRML / X3D files) will benefit from this, since they use this fixed-function pipeline shading model and can hence easily be mapped to glTF.

**Comment by tparisi** [comment]
Well I would also suggest "constant" (or "unlit") as well.

So you're suggesting skip Lambert, make Blinn-Phong be a single lighting model (which equation?). 

2 is better than 4.

:-)

**Comment by tparisi** [comment]
@fabrice I'm trying to understand how and when you generate the detail passes. In the Virtual City model for example I see 7 different shaders with a "Blinn" `lightingModel`. Why would you have seven? Are the actual GLSL shaders different? If so, how do they relate to a Blinn lighting model? Are you treating Blinn as a fallback?

**Comment by pjcozzi** [comment]
I'm fine with supporting fewer lighting models, but just a heads up that the hard part is different light types and multiple lights.  So supporting fewer lighting models is OK, but there is still a lot of work to generate shaders for all the lights, and the spec language needs to allow both forward and deferred shading, and the use of lights in the app's scene.

I _almost_ want to suggest keeping custom techniques as core spec, and then making this basic lighting model an extension that could go into spec in 1.1, for example, if it is widely used.

Here's a material without the extension:
```javascript
    "materials": {
            "material-id": {
                    "technique" : "technique1",
                    "values": {
                        "diffuse": [0.7,0.7,0.7,1],
                        "shininess": 38.4,
                        "specular": [0.2,0.2,0.2,1]
                    }
            }
        }
```
and with it:
```javascript
    "materials": {
            "material-id": {
                    "technique" : "this-can-be-used-or-ignored-when-the-extension-is-used",
                    "extensions": {
                        "KHR_extension_name" : {
                            "technique" : "LAMBERT", // naming convention similiar to glTF semantics
                            "values": {
                                "diffuse": [0.5,0.5,0.5,1]
                            }
                        }
                    }
            }
        }
```

**Comment by RemiArnaud** [comment]
Why not jumping ahead and use PBR materials definitions?
Nowadays its getting quite standard for many engine to offer this type of rendering.

Here's a good doc from unreal
https://docs.unrealengine.com/latest/INT/Engine/Rendering/Materials/PhysicallyBased/index.html
same type of material definition exists in Unity, Crytek, Valhalla ...

webGL engines would not have to be PBR rendering, could use a subset of the PBR data and do their best at some approximation. OpenGL engines can use the full PBR.

**Comment by mlimper** [comment]
> make Blinn-Phong be a single lighting model (which equation?).

Well, that probably needs some research, but here's a starting point, the X3D spec:
http://www.web3d.org/documents/specifications/19775-1/V3.2/Part01/components/lighting.html#Lightingequations

This verbose description includes different kinds and numbers of lights, fog, and so on.

It is using the halfway vector, so I thought that would make it "blinn-phong" (as opposed to "phong" using the reflected view vector instead)... not 100% sure about the naming though.

**Comment by mlimper** [comment]
> Why not jumping ahead and use PBR materials definitions?
> Nowadays its getting quite standard for many engine to offer this type of rendering.

Actually, we already wanted to propose an extension for this :-)

@thefranke has conducted a small study, investigating PBR parameters and their interpretation within the most common game engines

@mthoener you are currently working on this - could you provide an update?

**Comment by tparisi** [comment]
@RemiArnaud great idea and yes we've already been talking about it. 

I think the way to go is to put the common materials into an extension along the lines of what @pjcozzi proposed. It's well-understood and likely to be widely supported - but it wouldn't be required in the base space.

In the wake of doing the common materials, we can also start discussions about PBR. But those are going to take a little longer and we need to wrap up the base spec first.

There are also discussions underway about MDL, and also with Box (nee Verold) regarding other approaches to materials.

Long story short, an extension is the way to go, and FYI this means we will be pulling out lights, too. @fabrobinet wins! ;->

**Comment by RemiArnaud** [comment]
As far as webGL PBR, have you seen the marmoset viewer?

http://www.marmoset.co/viewer

— Remi

**Comment by tparisi** [comment]
Yes! They demoed it at our SIGGRAPH WebGL BoF and we were all really blown away. Amazing

**Comment by RemiArnaud** [comment]
That’s the level of rendering our artists are looking for.

Use case: an asset warehouse viewer for web browser.
Until then, we have to use the engine to render.

Go glTF !

— Remi

**Comment by tparisi** [comment]
@pjcozzi So here's my proposal for how to take both common materials and lighting out of the base spec and move them into an extension. We'll call the putative extension Basic Materials and give it the name "KHR_materials_basic"

It's pretty easy, actually. 3 changes:

1. Move the list of lights to the top-level `extensions` property 
2. Each material that uses a common technique has an `extensions` property with the technique name and values
3. Nodes also have an `extensions` property, where each light is placed (lights need to be part of the scene graph to be transformed etc)

I believe this is all we need. Please check out the example here.


```javascript
    "extensions": {

        "KHR_materials_basic" : {

            "lights": {
                "light1": {
                    "directional": {
                        "color": [
                            1,
                            1,
                            1
                        ]
                    },
                    "type": "directional"
                }
            },

        }
    },

    "materials": {
            "lambert1": {
                "extensions": {
                    "KHR_materials_basic" : {
                        // one of CONSTANT, BLINN, PHONG, LAMBERT
                        "technique" : "LAMBERT",
                        "values": {
                            "diffuse": [0.5,0.5,0.5,1]
                        }
                    }
                }
            }
        },

    "nodes": {
        "node1" : {
            "children" : [
            ],
            "extensions": {
                "KHR_materials_basic" : {
                    "light" : "light1",
                }
            }            
        }
    }
```

**Comment by tparisi** [comment]
@pjcozzi Now for the more brief version of techniques. Would love your reaction (and anyone else following this issue!)

Highlights:

1. This removes `pass` and `passes`. As you and I discussed, since we have no multi-pass in V1 let's not make an attempt at forward compatibility. We'll get it wrong the first time out, so why bother?
2. This takes out `instanceProgram`. The net result is collapsed by three levels: no `passes.defaultPass.instanceProgram`. Now attributes, program and uniforms are all at the same level as parameters. 
3. Also, without passes, `states` is at the top level of the technique.

Here is an example:

```javascript
    "materials": {
            "material1": {
                "technique" : "technique1",
                "values": {
                    "diffuse": [0.7,0.7,0.7,1],
                    "shininess": 38.4,
                    "specular": [0.2,0.2,0.2,1]
                }
            }
        },

    "techniques": {
        "technique1": {

            "parameters": {
                // Parameters details omitted for brevity; they haven't changed.
            },
            "attributes": {
                "a_normal": "normal",
                "a_position": "position",
                "a_texcoord0": "texcoord0"
            },
            "program": "program_0",
            "uniforms": {
                "u_ambient": "ambient",
                "u_diffuse": "diffuse",
                "u_emission": "emission",
                "u_light0Color": "light0Color",
                "u_light0Transform": "light0Transform",
                "u_modelViewMatrix": "modelViewMatrix",
                "u_normalMatrix": "normalMatrix",
                "u_projectionMatrix": "projectionMatrix",
                "u_shininess": "shininess",
                "u_specular": "specular"
            },
            "states": {
                "enable": [
                    2884,
                    2929
                ]
            }
        }
    }
```

**Comment by mlimper** [comment]
One question, if techniques are the default, I assume the "technique" property of "material" will be required. How do we deal with this when defining a simple material, using the proposed extension? AFAIK we agreed that extension may not define previously required parameters as optional.

**Comment by pjcozzi** [comment]
> @pjcozzi Now for the more brief version of techniques. Would love your reaction (and anyone else following this issue!)

All three changes are good with me.  Let's do it! :+1:

**Comment by pjcozzi** [comment]
> @pjcozzi So here's my proposal for how to take both common materials and lighting out of the base spec and move them into an extension.

Overall looks good.  We'll need to nail down the specifics as I mentioned above: interaction with `values`, textures, etc.

> KHR_materials_basic

I don't think `_basic` is the right name, but we can figure that out as we evolve the extension's spec, e.g., if it only did one lighting model, we could name it that.

> One question, if techniques are the default, I assume the "technique" property of "material" will be required. How do we deal with this when defining a simple material, using the proposed extension? AFAIK we agreed that extension may not define previously required parameters as optional.

Perhaps we write the glTF spec so it says `material.technique` is required unless an extension with a technique is provided (will need more precise language).  This shouldn't cause any more fragmentation  since the runtime needs to check `extensionsUsed` anyway.  When KHR_materials_basic is used, we could say the shader (if also included) takes precedence or vice versa.

Also, for nodes, we already allow "empty" nodes so including the light in the node's extension is perfect.

**Comment by tparisi** [comment]
Or, we just make the technique optional and allow an implementation to light the mesh any way it wants.

**Comment by pjcozzi** [comment]
> Or, we just make the technique optional and allow an implementation to light the mesh any way it wants.

I think this would be harder for a client to implement and would lead to inconsistencies among clients.

Is "required unless an extension overrides it" something we should avoid?

**Comment by emackey** [comment]
@pjcozzi I would suggest yes for the sake of clients that don't offer the extension.  My $0.02 would be, make it fully required, even if the value is ignored by clients that accept the extension, so the value is there as a fallback for less capable clients.

**Comment by tparisi** [comment]
@emackey My problem with this is that it puts more of a burden on exporters to generate good glTF. Not sure how big of a deal that is. Second smaller problem is file size and payload because then you have to define the techniques (a bit verbose but not that big) and send the GLSL code too

**Comment by pjcozzi** [comment]
> @emackey My problem with this is that it puts more of a burden on exporters to generate good glTF.

Exactly.  This is a lot of motivation, see #411.

**Comment by emackey** [comment]
OK.  Just seems like a circular dependency to me to have a piece of core spec be "required" or "not required" based on rules contained in a possibly unknown extension.  If "required" is not practical, maybe make it "not required" with an implied default unlit shader `vec4(1.0)` (or whatever is simplest and makes sense).

**Comment by pjcozzi** [comment]
> maybe make it "not required" with an implied default unlit shader vec4(1.0) (or whatever is simplest and makes sense).

This is what @tparisi suggested-ish, but it creates a burden for the client.  Now all clients need to support this even if it isn't useful in practice because all real assets either (1) have a technique, or (2) have a material defined in an extension.

**Comment by emackey** [comment]
The part of that suggestion I disagreed with was "any way it wants."  Perhaps the parameter is listed as optional, but the default behavior is to not render the mesh at all unless a technique or a material is defined.

**Comment by tparisi** [comment]
The way to make the technique optional is to define defaults. I don't see the harm in a default vertex shader behavior of modelview space and a default fragment shader behavior of "constant" lighting (i.e. no lighting). I don't see this as a huge burden for an importer implementation.

**Comment by tparisi** [comment]
Oh and I suppose we would have to define an emissive color to go with this if vertex colors aren't present. A typical default is half-gray.

**Comment by pjcozzi** [comment]
Sounds good.  I think we can make this trivial for runtime engines by defining the default using glTF - we need to define the parameters, shaders, etc., anyway.  Then we can then provide a utility function that adds the default to a glTF JSON object when needed (the Cesium team can write this) like we already do for other defaults: https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/ThirdParty/gltfDefaults.js

**Comment by tparisi** [comment]
THAT is brilliant.

@emackey thoughts?

Tony

**Comment by emackey** [comment]
Sounds great!  :+1:

**Comment by tparisi** [comment]
OK I'll work on the language for this plus a fully fleshed out basic materials spec... watch this space.

**Comment by pjcozzi** [comment]
Updated the schema based on https://github.com/KhronosGroup/glTF/issues/84#issuecomment-142185519 in #419

**Comment by pjcozzi** [comment]
@tparisi any update here?

**Comment by tparisi** [comment]
No I won't have time till tomorrow... stay tuned.

On Thu, Oct 1, 2015 at 5:57 AM, Patrick Cozzi <notifications@github.com>
wrote:

> @tparisi <https://github.com/tparisi> any update here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/KhronosGroup/glTF/issues/84#issuecomment-144719371>.
>



-- 


Tony Parisi                             tparisi@gmail.com
Follow me on Twitter!             http://twitter.com/auradeluxe
Read my blog at                     http://www.tonyparisi.com/
Learn WebGL                         http://learningwebgl.com/
Mobile                                    415.902.8002
Skype                                     auradeluxe

Read my books!
*Learning Virtual Reality*
http://www.amazon.com/Learning-Virtual-Reality-Experiences-Applications/dp/1491922834


*Programming 3D Applications in HTML5 and
WebGLhttp://www.amazon.com/Programming-Applications-HTML5-WebGL-Visualization/dp/1449362966
<http://www.amazon.com/Programming-Applications-HTML5-WebGL-Visualization/dp/1449362966>WebGL,
Up and Running*
http://www.amazon.com/dp/144932357X

**Comment by pjcozzi** [comment]
@tparisi can you post your update?

**Comment by tparisi** [comment]
Posted simplified techniques in 3094a39

Working on common materials in 2b2b3cc and beyond... need to complete each material/light definition but the main spec prose is there.

**Comment by pjcozzi** [comment]
Good start!

> This extension defines four light types: diffuse, point, spot, and ambient

"directional", not "diffuse", right?

> for Blinn, Blinn-Phong, Lamber and Unlit 

"Lamber" -> "Lambert"

> for Blinn, Blinn-Phong, ...

I'm not a lighting model expert, but I believe we want "Phong" and "Blinn-Phong" here, where Blinn-Phong uses the half vector for the specular component.  For example: http://ruh.li/GraphicsPhongBlinnPhong.html

**Comment by pjcozzi** [comment]
> Posted simplified techniques in 3094a39

Looks OK, but we still need to define the default.  See https://github.com/KhronosGroup/glTF/issues/84#issuecomment-143895190

**Comment by tparisi** [comment]
right on all counts, fixed typos and yes I need to define default.

**Comment by tparisi** [comment]
@pjcozzi oh and I think that means that `technique` is optional in a material?

**Comment by pjcozzi** [comment]
Yes, `technique` is optional.

**Comment by tparisi** [comment]
@pjcozzi can you review the common materials extension? Fairly done https://github.com/KhronosGroup/glTF/blob/spec-1.0/extensions/Khronos/KHR_materials_common/README.md

**Comment by RemiArnaud** [comment]
Just wondering how per vertex color would fit in this model ?

Regards
— Remi

**Comment by tparisi** [comment]
which model? common materials? yeah we probably need to handle per-vertex color and normal flags in there. good catch.

**Comment by tparisi** [comment]
@pjcozzi I need help figuring out rendering default if technique is optional. I mentioned 80% gray as a default. is that emissive or diffuse? e.g. if the implementation has lights it can do a diffuse color on that assuming normals are supplied (or generated). that's a lot more interesting than just emissive, which will look like indistinct blobs on the screen... requires more thought. feeling like opening a separate issue on this.

**Comment by pjcozzi** [comment]
Comments:

* The JSON examples are inconsistent.  Both `1` and `1.0` are used.  Arrays are sometimes on the same line and sometimes one element per line.
* `index_of_refraction` -> `indexOfRefraction`.  glTF uses camel case for property names.
   * Same for `constant_attenuation`, `linear_attenuation`, `falloff_angle`, and `falloff_exponent`.
* Are we sure `CONSTANT` needs all those properties?  How is each used?
* `FLOAT_VEC4` or `SAMPLER_2D` -> In glTF, what is sampler 2D?  Shouldn't this be a string (texture id)?  We need to precisely define where the texture coordinates come from, e.g., if both `diffuse` and `specular` are texures, do they use the same or different set of texture coordinates?
* `technique` - why does this have a default?  Perhaps it should be required.  We should also make this separate from the table and say the table is for the `values` object.
* `"relective": "texture_envmap"` - glTF does not support cube maps.
* This needs to include the equations.
* This needs to include some discussion on how vertex attributes use semantics to map material inputs, e.g., normals, texture coordinates, etc.

> To-do: Patrick need schema for each of the material types... can you help?

Maybe, but I did to finish the core spec reference doc first.  Will let you know later this week.

**Comment by pjcozzi** [comment]
> which model? common materials? yeah we probably need to handle per-vertex color and normal flags in there. good catch.

Or just use raw glTF techniques.  I don't know that this is widely enough used to justify making the spec more complicated.

**Comment by pjcozzi** [comment]
> @pjcozzi I need help figuring out rendering default if technique is optional. I mentioned 80% gray as a default. is that emissive or diffuse?

I would think `diffuse`, but it would also need some ambient.  Why not define it as a Lambert material and then we can fine tune the values when @tfili has an implementation.

**Comment by tparisi** [comment]
yeah that's a good idea.

I'm going to open another issue for the common materials extension. too much noise in here.

**Comment by tparisi** [comment]
moved common materials discussion to #424

**Comment by emackey** [comment]
> > @pjcozzi I need help figuring out rendering default if technique is optional. I mentioned 80% gray as a default. is that emissive or diffuse?
>
> I would think diffuse, but it would also need some ambient. Why not define it as a Lambert material and then we can fine tune the values when @tfili has an implementation.

Just my $0.02, but if the only reason we need a default is so that this property can be optional for the sake of extensions, then I think it's reasonable to expect most users to either supply the property or implement an extension.  So the default can be brain-dead, such as half-gray emmissive with no lighting, to keep it simple and to encourage supplying of better values.

**Comment by pjcozzi** [comment]
Good point @emackey.  That would be fine with me.

**Comment by tparisi** [comment]
@pjcozzi please review the language...

**Comment by pjcozzi** [comment]
Looks good with two comments:

* [ ] Perhaps add a short non-normative section on why, e.g., we don't expect apps to use this fallback, but instead want to allow materials in extensions without having to provide a dummy technique.
* [ ] Add glTF material/technique/program/shader example for the default shade.  @tfili can you provide this?

**Comment by pjcozzi** [comment]
@tfili do you happen to have this?

>  Add glTF material/technique/program/shader example for the default shade.

**Comment by pjcozzi** [comment]
Labeled `post 1.0`.

**Comment by pjcozzi** [comment]
@tfili this is all we need to close this:

> Add glTF material/technique/program/shader example for the default shader. @tfili can you provide this?

Can you provide this?

**Comment by tfili** [comment]
A default material would be a constant gray color, like defined with the material below using the KHR_materials_common extension.

```json
"materials": {
        "Effect1": {
            "extensions": {
                "KHR_materials_common": {
                    "technique": "CONSTANT",
                    "values": {
                        "emission": [
                            0.8,
                            0.8,
                            0.8,
                            1
                        ]
                    }
                }
            }
        }
    },
```

During runtime we would generate the following json objects

```json
    "materials": {
        "Effect1": {
            "values": {
                "emission": [
                    0.8,
                    0.8,
                    0.8,
                    1
                ]
            },
            "technique": "technique0"
        }
    },
    "programs": {
        "program0": {
            "attributes": [
                "a_normal",
                "a_position",
                "a_texcoord_0"
            ],
            "fragmentShader": "fragmentShader0",
            "vertexShader": "vertexShader0"
        }
    },
    "shaders": {
        "vertexShader0": {
            "type": 35633,
            "uri": "data:text/plain;base64,...",
        },
        "fragmentShader0": {
            "type": 35632,
            "uri": "data:text/plain;base64,..."
        }
    },
    "techniques": {
        "technique0": {
            "attributes": {
                "a_normal": "normal",
                "a_position": "position"
            },
            "parameters": {
                "modelViewMatrix": {
                    "semantic": "MODELVIEW",
                    "type": 35676
                },
                "normalMatrix": {
                    "semantic": "MODELVIEWINVERSETRANSPOSE",
                    "type": 35675
                },
                "projectionMatrix": {
                    "semantic": "PROJECTION",
                    "type": 35676
                },
                "emission": {
                    "type": 35666
                },
                "normal": {
                    "semantic": "NORMAL",
                    "type": 35665
                },
                "position": {
                    "semantic": "POSITION",
                    "type": 35665
                }
            },
            "program": "program0",
            "states": {
                "enable": [
                    2884,
                    2929
                ]
            },
            "uniforms": {
                "u_modelViewMatrix": "modelViewMatrix",
                "u_normalMatrix": "normalMatrix",
                "u_projectionMatrix": "projectionMatrix",
                "u_emission": "emission"
            }
        }
    }
```

The shaders would be

Vertex:
```glsl
precision highp float;
uniform mat4 u_modelViewMatrix;
uniform mat3 u_normalMatrix;
uniform mat4 u_projectionMatrix;
attribute vec3 a_normal;
attribute vec3 a_position;
void main(void) {
  gl_Position = u_projectionMatrix * u_modelViewMatrix * vec4(a_position,1.0);
}
```

Fragment:
```glsl
precision highp float;
uniform vec4 u_emission;
void main(void) {
  gl_FragColor = u_emission;
}
```

Obviously encoding in a data URI wouldn't be the ideal situation but it would work with existing engines. One option could be to add an `extras` property the contains the glsl source string directly, but that would be up to the implementer.

**Comment by pjcozzi** [comment]
Thanks @tfili.  I added this to the spec in the `1.0-final` branch. fefc1526c74291afd673501b4568af51513debb3
